### PR TITLE
Ceres FATE extensions

### PR DESCRIPTION
**Pull request by hanssv**
This is a set of extensions to FATE for the Ceres protocol upgrade. 

In this PR we introduce VM_SOPHIA_FATE_3 - the third version of the FATE VM - all the new operations (bitwise operations on integers, MULMOD, addr_to_bytes and poseidon hash) are restricted to the new VM version. And likewise the new VM version is only available in the Ceres protocol.

 This PR is supported by the Æternity Crypto Foundation

### CODE REVIEW

File: apps/aecontract/test/aect_test_utils.erl [hunk 1]
```diff
@@ -414,12 +415,13 @@ legacy_compile(Vsn, SrcFile) ->
...
-        ?SOPHIA_LIMA_AEVM -> filename:join([BaseDir, "v4.3.1", "aesophia_cli"]) ++ " --backend=aevm";
-        ?SOPHIA_LIMA_FATE -> filename:join([BaseDir, "v4.3.1", "aesophia_cli"]);
-        ?SOPHIA_IRIS_FATE -> filename:join([BaseDir, "v6.1.0", "aesophia_cli"])
+        ?SOPHIA_ROMA       -> filename:join([BaseDir, "v1.4.0", "aesophia_cli"]);
+        ?SOPHIA_MINERVA    -> filename:join([BaseDir, "v2.1.0", "aesophia_cli"]);
+        ?SOPHIA_FORTUNA    -> filename:join([BaseDir, "v3.2.0", "aesophia_cli"]);
+        ?SOPHIA_LIMA_AEVM  -> filename:join([BaseDir, "v4.3.1", "aesophia_cli"]) ++ " --backend=aevm";
+        ?SOPHIA_LIMA_FATE  -> filename:join([BaseDir, "v4.3.1", "aesophia_cli"]);
+        ?SOPHIA_IRIS_FATE  -> filename:join([BaseDir, "v6.1.0", "aesophia_cli"]);
+        ?SOPHIA_CERES_FATE -> filename:join([BaseDir, "v6.1.0", "aesophia_cli"])
```
Comment (uwiger) [inline apps/aecontract/test/aect_test_utils.erl:424]: Not v7.0.0?

File: apps/aeutils/priv/generate_poseidon.sh [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+CURVE="BLS12_381"
+PRIME="0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001"
+SIZE=255
+
+# CURVE="BN128"
+# PRIME="0x30644e72e131a029b85045b68181585d2833e84879b9709143e1f593f0000001"
+# SIZE=254
+
+SCRIPT=/Users/hans/Repos/hadeshash/code/generate_parameters_grain.sage
```
Comment (uwiger) [inline apps/aeutils/priv/generate_poseidon.sh:15]: Not a very portable path?

**Review by uwiger — APPROVED** [review]

File: apps/aecontract/test/aect_test_utils.erl (hunk 1 above)
Comment (hanssv) [reply apps/aecontract/test/aect_test_utils.erl:424]: There is no v7.0.0 of the CLI compiler yet, and Ceres shouldn't need the CLI compiler either - my guess is that there is a poorly written test case somewhere that does the wrong thing... I'll check.

File: apps/aeutils/priv/generate_poseidon.sh (hunk 2 above)
Comment (hanssv) [reply apps/aeutils/priv/generate_poseidon.sh:15]: Nah, let me replace that with a `git clone` - including the script is mainly for reference I don't expect anyone to run it 😅

**Review by uwiger — DISMISSED** [review]

**Review by uwiger — DISMISSED** [review]

File: apps/aecontract/test/aect_test_utils.erl (hunk 1 above)
Comment (hanssv) [reply apps/aecontract/test/aect_test_utils.erl]: The caching/aci seems to use this under some circumstances... I won't try to untangle that mess now...

**Review by uwiger — APPROVED** [review]

File: apps/aefate/src/aefa_fate_op.erl [hunk 3]
```diff
@@ -2675,6 +2764,12 @@ op(Op, Arg1, Arg2) ->
     aefa_fate:abort({type_error, Op, [Arg1, Arg2]}).
 
 %% Terinay operations
```
Comment (nikita-fuchs) [inline apps/aefate/src/aefa_fate_op.erl:2766]: *ternary

Comment (hanssv) [reply apps/aefate/src/aefa_fate_op.erl:2766]: I'm pretty sure I didn't write that 🤔 

Ah, not part of the change, but yes correct.

**Review by velzevur — APPROVED** [review]

**Review by velzevur — DISMISSED** [review]

**Review by uwiger — DISMISSED** [review]

**Review by seanhinde — DISMISSED** [review]
Very nice

**Review by velzevur — DISMISSED** [review]

**Review by velzevur — APPROVED** [review]

**Review by seanhinde — APPROVED** [review]

**Pull request merged**